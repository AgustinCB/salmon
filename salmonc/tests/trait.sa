trait Test {
    test(a);
    class staticmethod();
}

class TestImpl {
    pepe() {}
}

trait Test for TestImpl {
    test(a) {
        print a;
    }
    class staticmethod() {
        print "static";
    }
}

var obj = TestImpl();
obj.test("method");
TestImpl.staticmethod();
print Test istype Trait;
print TestImpl istype Class;
print TestImpl istype Test;
print obj istype TestImpl;
