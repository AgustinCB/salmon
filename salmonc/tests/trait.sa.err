Constants: [SimpleValue(Function { ip: 171, arity: 1, uplifts: None }), SimpleValue(Function { ip: 171, arity: 2, uplifts: None }), SimpleValue(Integer(1)), SimpleValue(Integer(4)), SimpleValue(Integer(3)), SimpleValue(String(0)), SimpleValue(Nil), SimpleValue(Function { ip: 185, arity: 1, uplifts: None }), SimpleValue(String(1)), SimpleValue(String(7)), SimpleValue(String(11)), SimpleValue(String(31)), SimpleValue(Integer(0)), SimpleValue(String(37)), SimpleValue(String(47)), SimpleValue(Object { address: 51, tags: 211 }), SimpleValue(String(218)), SimpleValue(String(225)), SimpleValue(String(231)), SimpleValue(String(241)), SimpleValue(String(251)), SimpleValue(String(263)), SimpleValue(String(270)), SimpleValue(String(276)), SimpleValue(String(288)), SimpleValue(String(300)), SimpleValue(String(320))]
Instructions: [Instruction { instruction_type: Constant(0), location: 0 }, Instruction { instruction_type: SetGlobal(1), location: 0 }, Instruction { instruction_type: Pop, location: 0 }, Instruction { instruction_type: Constant(1), location: 2 }, Instruction { instruction_type: SetGlobal(2), location: 2 }, Instruction { instruction_type: Pop, location: 2 }, Instruction { instruction_type: Constant(7), location: 3 }, Instruction { instruction_type: SetGlobal(3), location: 3 }, Instruction { instruction_type: Pop, location: 3 }, Instruction { instruction_type: Constant(11), location: 4 }, Instruction { instruction_type: Constant(12), location: 4 }, Instruction { instruction_type: ObjectAlloc, location: 4 }, Instruction { instruction_type: AddTag, location: 4 }, Instruction { instruction_type: SetGlobal(4), location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: Constant(16), location: 4 }, Instruction { instruction_type: Constant(17), location: 4 }, Instruction { instruction_type: Constant(15), location: 4 }, Instruction { instruction_type: AddTag, location: 4 }, Instruction { instruction_type: AddTag, location: 4 }, Instruction { instruction_type: SetGlobal(5), location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: GetGlobal(2), location: 4 }, Instruction { instruction_type: Constant(9), location: 4 }, Instruction { instruction_type: GetGlobal(5), location: 4 }, Instruction { instruction_type: ObjectSet, location: 4 }, Instruction { instruction_type: Swap, location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: SetGlobal(5), location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: GetGlobal(3), location: 4 }, Instruction { instruction_type: Constant(10), location: 4 }, Instruction { instruction_type: GetGlobal(5), location: 4 }, Instruction { instruction_type: ObjectSet, location: 4 }, Instruction { instruction_type: Swap, location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: SetGlobal(5), location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: Constant(21), location: 4 }, Instruction { instruction_type: GetGlobal(5), location: 4 }, Instruction { instruction_type: AddTag, location: 4 }, Instruction { instruction_type: SetGlobal(5), location: 4 }, Instruction { instruction_type: Pop, location: 4 }, Instruction { instruction_type: GetGlobal(5), location: 5 }, Instruction { instruction_type: Duplicate, location: 5 }, Instruction { instruction_type: CheckType(7), location: 5 }, Instruction { instruction_type: JmpIfFalse(3), location: 5 }, Instruction { instruction_type: Constant(17), location: 5 }, Instruction { instruction_type: Swap, location: 5 }, Instruction { instruction_type: RemoveTag, location: 5 }, Instruction { instruction_type: Call, location: 5 }, Instruction { instruction_type: SetGlobal(0), location: 5 }, Instruction { instruction_type: Pop, location: 5 }, Instruction { instruction_type: Constant(22), location: 6 }, Instruction { instruction_type: GetGlobal(0), location: 6 }, Instruction { instruction_type: Duplicate, location: 6 }, Instruction { instruction_type: Constant(17), location: 6 }, Instruction { instruction_type: Swap, location: 6 }, Instruction { instruction_type: CheckTag, location: 6 }, Instruction { instruction_type: JmpIfFalse(4), location: 6 }, Instruction { instruction_type: Constant(23), location: 6 }, Instruction { instruction_type: Swap, location: 6 }, Instruction { instruction_type: ObjectGet, location: 6 }, Instruction { instruction_type: Jmp(12), location: 6 }, Instruction { instruction_type: Constant(24), location: 6 }, Instruction { instruction_type: Swap, location: 6 }, Instruction { instruction_type: ObjectHas, location: 6 }, Instruction { instruction_type: JmpIfFalse(5), location: 6 }, Instruction { instruction_type: Constant(24), location: 6 }, Instruction { instruction_type: Swap, location: 6 }, Instruction { instruction_type: ObjectGet, location: 6 }, Instruction { instruction_type: Call, location: 6 }, Instruction { instruction_type: Jmp(3), location: 6 }, Instruction { instruction_type: Constant(9), location: 6 }, Instruction { instruction_type: Swap, location: 6 }, Instruction { instruction_type: ObjectGet, location: 6 }, Instruction { instruction_type: Duplicate, location: 6 }, Instruction { instruction_type: CheckType(7), location: 6 }, Instruction { instruction_type: JmpIfFalse(3), location: 6 }, Instruction { instruction_type: Constant(17), location: 6 }, Instruction { instruction_type: Swap, location: 6 }, Instruction { instruction_type: RemoveTag, location: 6 }, Instruction { instruction_type: Call, location: 6 }, Instruction { instruction_type: Pop, location: 6 }, Instruction { instruction_type: GetGlobal(5), location: 7 }, Instruction { instruction_type: Duplicate, location: 7 }, Instruction { instruction_type: Constant(17), location: 7 }, Instruction { instruction_type: Swap, location: 7 }, Instruction { instruction_type: CheckTag, location: 7 }, Instruction { instruction_type: JmpIfFalse(4), location: 7 }, Instruction { instruction_type: Constant(10), location: 7 }, Instruction { instruction_type: Swap, location: 7 }, Instruction { instruction_type: ObjectGet, location: 7 }, Instruction { instruction_type: Jmp(12), location: 7 }, Instruction { instruction_type: Constant(25), location: 7 }, Instruction { instruction_type: Swap, location: 7 }, Instruction { instruction_type: ObjectHas, location: 7 }, Instruction { instruction_type: JmpIfFalse(5), location: 7 }, Instruction { instruction_type: Constant(25), location: 7 }, Instruction { instruction_type: Swap, location: 7 }, Instruction { instruction_type: ObjectGet, location: 7 }, Instruction { instruction_type: Call, location: 7 }, Instruction { instruction_type: Jmp(3), location: 7 }, Instruction { instruction_type: Constant(20), location: 7 }, Instruction { instruction_type: Swap, location: 7 }, Instruction { instruction_type: ObjectGet, location: 7 }, Instruction { instruction_type: Duplicate, location: 7 }, Instruction { instruction_type: CheckType(7), location: 7 }, Instruction { instruction_type: JmpIfFalse(3), location: 7 }, Instruction { instruction_type: Constant(17), location: 7 }, Instruction { instruction_type: Swap, location: 7 }, Instruction { instruction_type: RemoveTag, location: 7 }, Instruction { instruction_type: Call, location: 7 }, Instruction { instruction_type: Pop, location: 7 }, Instruction { instruction_type: Constant(5), location: 8 }, Instruction { instruction_type: Constant(11), location: 8 }, Instruction { instruction_type: GetGlobal(4), location: 8 }, Instruction { instruction_type: CheckTag, location: 8 }, Instruction { instruction_type: ToStr, location: 8 }, Instruction { instruction_type: StringConcat, location: 8 }, Instruction { instruction_type: Push, location: 8 }, Instruction { instruction_type: Strlen, location: 8 }, Instruction { instruction_type: Swap, location: 8 }, Instruction { instruction_type: Constant(2), location: 8 }, Instruction { instruction_type: Constant(4), location: 8 }, Instruction { instruction_type: Constant(3), location: 8 }, Instruction { instruction_type: Syscall, location: 8 }, Instruction { instruction_type: Pop, location: 8 }, Instruction { instruction_type: Constant(5), location: 9 }, Instruction { instruction_type: Constant(17), location: 9 }, Instruction { instruction_type: GetGlobal(5), location: 9 }, Instruction { instruction_type: CheckTag, location: 9 }, Instruction { instruction_type: ToStr, location: 9 }, Instruction { instruction_type: StringConcat, location: 9 }, Instruction { instruction_type: Push, location: 9 }, Instruction { instruction_type: Strlen, location: 9 }, Instruction { instruction_type: Swap, location: 9 }, Instruction { instruction_type: Constant(2), location: 9 }, Instruction { instruction_type: Constant(4), location: 9 }, Instruction { instruction_type: Constant(3), location: 9 }, Instruction { instruction_type: Syscall, location: 9 }, Instruction { instruction_type: Pop, location: 9 }, Instruction { instruction_type: Constant(5), location: 10 }, Instruction { instruction_type: Constant(21), location: 10 }, Instruction { instruction_type: GetGlobal(5), location: 10 }, Instruction { instruction_type: CheckTag, location: 10 }, Instruction { instruction_type: ToStr, location: 10 }, Instruction { instruction_type: StringConcat, location: 10 }, Instruction { instruction_type: Push, location: 10 }, Instruction { instruction_type: Strlen, location: 10 }, Instruction { instruction_type: Swap, location: 10 }, Instruction { instruction_type: Constant(2), location: 10 }, Instruction { instruction_type: Constant(4), location: 10 }, Instruction { instruction_type: Constant(3), location: 10 }, Instruction { instruction_type: Syscall, location: 10 }, Instruction { instruction_type: Pop, location: 10 }, Instruction { instruction_type: Constant(5), location: 11 }, Instruction { instruction_type: Constant(16), location: 11 }, Instruction { instruction_type: GetGlobal(0), location: 11 }, Instruction { instruction_type: CheckTag, location: 11 }, Instruction { instruction_type: ToStr, location: 11 }, Instruction { instruction_type: StringConcat, location: 11 }, Instruction { instruction_type: Push, location: 11 }, Instruction { instruction_type: Strlen, location: 11 }, Instruction { instruction_type: Swap, location: 11 }, Instruction { instruction_type: Constant(2), location: 11 }, Instruction { instruction_type: Constant(4), location: 11 }, Instruction { instruction_type: Constant(3), location: 11 }, Instruction { instruction_type: Syscall, location: 11 }, Instruction { instruction_type: Pop, location: 11 }, Instruction { instruction_type: Return, location: 11 }, Instruction { instruction_type: Constant(5), location: 2 }, Instruction { instruction_type: GetLocal(1), location: 2 }, Instruction { instruction_type: ToStr, location: 2 }, Instruction { instruction_type: StringConcat, location: 2 }, Instruction { instruction_type: Push, location: 2 }, Instruction { instruction_type: Strlen, location: 2 }, Instruction { instruction_type: Swap, location: 2 }, Instruction { instruction_type: Constant(2), location: 2 }, Instruction { instruction_type: Constant(4), location: 2 }, Instruction { instruction_type: Constant(3), location: 2 }, Instruction { instruction_type: Syscall, location: 2 }, Instruction { instruction_type: Pop, location: 2 }, Instruction { instruction_type: Constant(6), location: 2 }, Instruction { instruction_type: Return, location: 2 }, Instruction { instruction_type: Constant(5), location: 3 }, Instruction { instruction_type: Constant(8), location: 3 }, Instruction { instruction_type: ToStr, location: 3 }, Instruction { instruction_type: StringConcat, location: 3 }, Instruction { instruction_type: Push, location: 3 }, Instruction { instruction_type: Strlen, location: 3 }, Instruction { instruction_type: Swap, location: 3 }, Instruction { instruction_type: Constant(2), location: 3 }, Instruction { instruction_type: Constant(4), location: 3 }, Instruction { instruction_type: Constant(3), location: 3 }, Instruction { instruction_type: Syscall, location: 3 }, Instruction { instruction_type: Pop, location: 3 }, Instruction { instruction_type: Constant(6), location: 3 }, Instruction { instruction_type: Return, location: 3 }]
Locations: [Location { address: 320, line: 5 }, Location { address: 320, line: 9 }, Location { address: 320, line: 11 }, Location { address: 320, line: 14 }, Location { address: 320, line: 0 }, Location { address: 320, line: 18 }, Location { address: 320, line: 19 }, Location { address: 320, line: 20 }, Location { address: 320, line: 21 }, Location { address: 320, line: 22 }, Location { address: 320, line: 23 }, Location { address: 320, line: 24 }]
