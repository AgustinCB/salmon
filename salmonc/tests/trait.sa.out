method
static
true
true
true
true
